dev-docs/best-practices/microservices.md

```diff
@@ -0,0 +1,111 @@
+# Building Microservices
+
+All Taskcluster services are implemented in subdirectories of the the `services/` directory in this respository.
+
+The Taskcluster team shares responsibility for all services, although people may "specialize" in specific services and have additional expertise there.
+This shared responsibility is easier for everyone if the implementations are similar, avoiding surprises when moving from one service to another.
+This document aims to collect the practices and standards we've agreed on.
+
+These conventions are strongly encouraged for new services and contributions updating existing services to follow them are always welcome.
+When we have a good reason to not follow the convention for a specific service, we document why.
+
+## Independence
+
+Although they ar developed in the same repository, each service must operate as an independent entity, using only the public interfaces of other services.
+Package imports between services are strictly forbidden.
+Where this seems like a good idea, prefer to create a new [library](libraries) instead.
+
+## Package Mechanics
+
+A service has a `serviceName` and a `projectName`, the latter often having a `taskclsuter-` prefix.
+For example, `auth` and `taskcluster-auth`, respectively.
+
+A service is implemented in `services/<serviceName>`.
+That directory should have a `package.json` containing `name: "<projectName>"`.
+Dependencies go in this `package.json`, while dev dependencies go in the root `package.json`.
+
+Source code should be in `src/`.
+No transpilation should be used: write JS that can be interpreted directly by the Node version in use in the repository.
+(The web-server service is an exception, since it uses webpack to load `.graphql` files)
+
+## Implementation
+
+### taskcluster-lib-loader
+
+The main entry-point for the service should be a file called `main.js`, which should use [taskcluster-lib-loader](../../libraries/loader) for loading components.
+
+### taskcluster-lib-api
+
+The API definition should be in a file called `api.js`:
+
+```js
+var api = new API({
+  // ...
+});
+
+// Export api
+module.exports = api;
+
+/** Get hook groups **/
+api.declare({
+  // ...
+});
+// ...
+```
+
+This is then imported and set up in `main.js`:
+
+```js
+{
+  router: {
+    requires: ['cfg', 'profile', 'validator', 'monitor'],
+    setup: ({cfg, profile, validator, monitor}) => {
+      return v1.setup({
+        context: {},
+        authBaseUrl:      cfg.taskcluster.authBaseUrl,
+        publish:          profile === 'production',
+        baseUrl:          cfg.server.publicUrl + '/v1',
+        referencePrefix:  'myservice/v1/api.json',
+        aws:              cfg.aws,
+        validator,
+        monitor,
+      });
+    },
+  },
+}
+```
+
+#### Error Handling
+
+Do not use `res.status(..)` to return error messages.
```

Comment:
Should we disable using res.status programmatically? Either could be some sort of lint or overriding it somehow.

Reply:
It seems we only use `res.status` in 6 service files at the moment.

Reply:
There are legitimate uses (like returning a response code without a body).  It's just *error* responses that should not..